### DISCUSSION TITLE
a run is active - exception (c#, open AI assistant)

### DISCUSSION THREAD

**Discussion by JP-droidic** [microsoft/semantic-kernel#9633] [Q&A] [answered]
Hi team,

not sure if anyone else has noticed this, but we are getting more and more often the following exceptions:

```
System.ClientModel.ClientResultException: HTTP 400 (invalid_request_error: ) 
Can't add messages to thread_8vvUmiYP5yNpmkwL2DF01Usj while a run run_xq3NOfM700rAyoJJ6tINKzFM is active. 
at OpenAI.ClientPipelineExtensions.ProcessMessageAsync(ClientPipeline pipeline, PipelineMessage message, RequestOptions options) 
at OpenAI.Assistants.InternalAssistantMessageClient.CreateMessageAsync(String threadId, BinaryContent content, RequestOptions options) 
at OpenAI.Assistants.AssistantClient.CreateMessageAsync(String threadId, MessageRole role, IEnumerable`1 content, MessageCreationOptions options, CancellationToken cancellationToken) 
at Microsoft.SemanticKernel.Agents.OpenAI.Internal.AssistantThreadActions.CreateMessageAsync(AssistantClient client, String threadId, ChatMessageContent message, CancellationToken cancellationToken) ]
at ...
```
The scenario is,

a user asks a question (new chat), we create an assistant and start a thread.
using streaming result is shown to the user 
user asks a second question  
we get the error message

This is completely random, I would say one in 7 chats gets the error message

**Comment by crickman** [comment, marked as answer]
HI @JP-droidic - On the surface this seems to indicate the perhaps the user message is being added while the run is still active.  I might double check any potential concurrency issues and also verify that the all of the (`IAsyncEnumerable`) streamed responses have been processed.

Also, can you confirm whether you are targeting the Open AI endpoint or an Azure AI endpoint?

In the meantime, I'll setup a local stress test to explore this dynamic further.  If I am able to reproduce this locally, it will aid in understanding what may be at play here.

**Comment by JP-droidic** [comment]
It is very strange indeed.  LLM stopped returning results(stream) maybe even a minute after the second message was sent. We use Open AI endpoint.

Thank!

**Comment by crickman** [comment]
Review the API docs and saw this confirmation on what is going on:

https://platform.openai.com/docs/assistants/deep-dive#thread-locks

The error message is pretty clear.  I suppose the question is why is the run active?  I reviewed the code and termination states.  My only thought on how a run might be active is if an exception was thrown; although, I'd expect such a failure to bubble up into your call-stack.  I'll keep looking, but can you also please examine logs to acertain if perhap an exception has occurred?  If so, this will help guide the fix.

**Reply by chesnelg** [reply]
I am experiencing the same error, and this is indeed after an exception occurred.
The Agent keeps on invoke the same Navite Plug-In and fails during parameter validation - a required parameter is not provided
Here is the error:

> fail: Microsoft.SemanticKernel.KernelFunction[0]
>       Function XXX-XXX failed. Error: Missing argument for function parameter 'XXX'
>       Microsoft.SemanticKernel.KernelException: Missing argument for function parameter 'XXX'
>        ---> System.ArgumentException: Missing argument for function parameter (Parameter 'XXX')
>          --- End of inner exception stack trace ---
>          at Microsoft.SemanticKernel.KernelFunctionFromMethod.<>c__DisplayClass23_0.<GetParameterMarshalerDelegate>g__parameterFunc|9(KernelFunction _, Kernel kernel, KernelArguments arguments, CancellationToken __)

I am using "Microsoft.SemanticKernel.Agents.OpenAI" Version="1.37.0-alpha"
Experiencing the pb with gtp-3.5 and gtp4

**Comment by sensation61** [comment]
Hi Team,

I'm getting the same error. Can you help me? Did you manage to solve the problem?

```
Package: Azure.AI.OpenAI
Package Version:  2.1.0
```
 
```
OperationCanceledException System.ClientModel.ClientResultException: HTTP 400 (invalid_request_error: )
Can't add messages to thread_GPsyTOSjKXGgVutlapRshnKY while a run run_3gDS5YDnhuHGlOoOqVIOuTQh is active.
         at Azure.AI.OpenAI.ClientPipelineExtensions.ProcessMessageAsync(ClientPipeline pipeline, PipelineMessage message, RequestOptions options)
         at Azure.AI.OpenAI.Assistants.AzureAssistantClient.CreateMessageAsync(String threadId, BinaryContent content, RequestOptions options)
         at OpenAI.Assistants.AssistantClient.CreateMessageAsync(String threadId, MessageRole role, IEnumerable`1 content, MessageCreationOptions options, CancellationToken cancellationToken)
```

**Comment by JP-droidic** [comment]
I "fixed" this by killing all active threads before running a new query 

```
var threads =  assistantClient.GetRuns(threadId).Where(a => a.Status == RunStatus.InProgress).ToList();
        if (threads.Any())
        {
            foreach (var el in threads)
            {
                await assistantClient.CancelRunAsync(threadId, el.Id);
            }
        }
```

**Comment by sensation61** [comment]
Thank you for sharing your solution with me. I implemented the code you provided, and it worked successfully! Killing all active threads before running a new query resolved the issue I was facing.
